### PR TITLE
Issue #2: Refactored the MapFragment, MapViewModel and MapRepository to better represent the MVVM structure

### PR DESCRIPTION
**Pull request by KhaledBadran**
#2 : This PR has the goal of refactoring the map and its related classes. The following changing were introduced with this PR:

- The initialization of the GoogleMap object has been majorly refactored to be done through a Map class (serves as the a model for the map) and the MapViewModel. This means that the MapFragment now only requests and already-initialized map object from the ViewModel.
- Logic in the MapFragment that searches and retrieves a building object using the map polygon has been moved to the MapViewModel as such logic more related to the responsibilities of the ViewModel in an MVVM structure. 
-  Reading from external files namely, buildings.json, has been moved from the MapViewModel to the MapRepository as its repository classes are responsible for handling external information.
- Added a unit test to the MapRepository that checks the newly refactored structure of that class, specifically, the existence of the external file.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -5,6 +5,8 @@ apply plugin: 'kotlin-android-extensions'
 apply plugin: 'kotlin-kapt'
 apply plugin: 'org.jetbrains.dokka'
 
+apply plugin: 'org.jetbrains.dokka'
+
```
Comment (Amanda-Kolopa) [inline app/build.gradle]: Why are you applying the same plugin twice?

File: app/src/main/java/com/droidhats/campuscompass/models/Map.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    var mapFragmentOnMarkerClickListener: GoogleMap.OnMarkerClickListener,
+    var mapFragmentOnPolygonClickListener: GoogleMap.OnPolygonClickListener,
+    var activity: FragmentActivity
+) {
+
+    companion object {
+        private const val MAP_PADDING_TOP = 200
+        private const val MAP_PADDING_RIGHT = 15
+    }
+
```
Comment (NicholasNagy) [inline app/src/main/java/com/droidhats/campuscompass/models/Map.kt]: why double space here?

File: app/src/main/java/com/droidhats/campuscompass/models/Map.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+     * Initializes the map, and attaches listeners to it.
+     */
+    init {
+        //updating map type we can choose between  4 types : MAP_TYPE_NORMAL, MAP_TYPE_SATELLITE, MAP_TYPE_TERRAIN, MAP_TYPE_HYBRID
+        googleMap.mapType = GoogleMap.MAP_TYPE_NORMAL
+
+        //initializing vars for get last current location
+        googleMap.uiSettings.isZoomControlsEnabled = true
+        googleMap.setOnMarkerClickListener(mapFragmentOnMarkerClickListener)
+
```
Comment (NicholasNagy) [inline app/src/main/java/com/droidhats/campuscompass/models/Map.kt]: why double space here?

File: app/src/main/java/com/droidhats/campuscompass/models/Map.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        if ((activity as MainActivity).checkLocationPermission()) {
+            //Enables the my-location layer which draws a light blue dot on the user’s location.
+            // It also adds a button to the map that, when tapped, centers the map on the user’s location.
+            googleMap.isMyLocationEnabled = true
+        }
+
+        googleMap.setOnPolygonClickListener(mapFragmentOnPolygonClickListener)
+    }
+
```
Comment (NicholasNagy) [inline app/src/main/java/com/droidhats/campuscompass/models/Map.kt]: why double space?

File: app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt [hunk 5]
```diff
@@ -1,24 +1,92 @@
...
-        campuses = MapRepository.getInstance(json).getCampuses()
+        campuses = MapRepository.getInstance(context).getCampuses()
+    }
+
+
+    /**
+     * Returns a list of campus objects (SJW and Loyola).
+     */
+    fun getCampuses(): List<Campus> = campuses!!
+
```
Comment (NicholasNagy) [inline app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt]: why double space here?

File: app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt [hunk 6]
```diff
@@ -1,24 +1,92 @@
...
+        for (campus in this.campuses!!) {
+            for (building in campus.getBuildings()) {
+                if (polygonTag == building.getName())
+                    selectedBuilding = building
+            }
+        }
+
+        return selectedBuilding
+    }
+
```
Comment (NicholasNagy) [inline app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt]: why all these empty lines?

File: app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt [hunk 7]
```diff
@@ -284,14 +256,9 @@ class MapFragment : Fragment(), OnMapReadyCallback, GoogleMap.OnMarkerClickListe
...
-            var selectedBuilding : Building? = null
-            for (campus in viewModel.getCampuses()) {
-                for (building in campus.getBuildings()) {
-                    if (p.tag.toString() == building.getName())
-                        selectedBuilding  = building
-                }
-            }
+            //Get the building object from the polygon that the user clicked on
+            var selectedBuilding : Building? = viewModel.findBuildingByPolygonTag(p.tag.toString())
+
```
Comment (NicholasNagy) [inline app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt]: why add this extra space?

File: app/src/test/java/com/droidhats/campuscompass/repositories/MapRepositoryTest.kt [hunk 8]
```diff
@@ -49,7 +53,27 @@ class MapRepositoryTest {
             "        ]\n" +
             "    }]\n" +
             "}"
-    private var instance: MapRepository = MapRepository.getInstance(json)
+
```
Comment (NicholasNagy) [inline app/src/test/java/com/droidhats/campuscompass/repositories/MapRepositoryTest.kt]: why add 2 empty lines?

File: app/build.gradle (hunk 1 above)
Comment (KhaledBadran) [reply app/build.gradle]: Thanks, I've removed the duplicate line in my latest commit.

File: app/src/main/java/com/droidhats/campuscompass/models/Map.kt (hunk 2 above)
Comment (KhaledBadran) [reply app/src/main/java/com/droidhats/campuscompass/models/Map.kt]: Thanks, I've removed the double space in my latest commit to this branch.

File: app/src/main/java/com/droidhats/campuscompass/models/Map.kt (hunk 3 above)
Comment (KhaledBadran) [reply app/src/main/java/com/droidhats/campuscompass/models/Map.kt]: Same here, I've removed the double space in my latest commit to this branch.

File: app/src/main/java/com/droidhats/campuscompass/models/Map.kt (hunk 4 above)
Comment (KhaledBadran) [reply app/src/main/java/com/droidhats/campuscompass/models/Map.kt]: The double space has been removed in my latest commit to this branch.

File: app/src/test/java/com/droidhats/campuscompass/repositories/MapRepositoryTest.kt (hunk 8 above)
Comment (KhaledBadran) [reply app/src/test/java/com/droidhats/campuscompass/repositories/MapRepositoryTest.kt]: Agreed, I've removed the extra line.

File: app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt (hunk 7 above)
Comment (KhaledBadran) [reply app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt]: I've removed the extra space in my new commit.

File: app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt (hunk 6 above)
Comment (KhaledBadran) [reply app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt]: The unnecessary lines were removed in the last commit.

File: app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt (hunk 5 above)
Comment (KhaledBadran) [reply app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt]: Removed the double space in the new commit.

**Pull request merged**